File: cua.py
import sys
import time
from evdev import ecodes, InputEvent, InputDevice, UInput, categorize

def special_keys_mapping(event):
    if event.code == ecodes.KEY_LEFTALT:
        return InputEvent(event.sec, event.usec, event.type, ecodes.KEY_LEFTCTRL, event.value)
    elif event.code == ecodes.KEY_LEFTCTRL:
        return InputEvent(event.sec, event.usec, event.type, ecodes.KEY_LEFTALT, event.value)
    else:
        return event

special_keys_pressed = {
    ecodes.KEY_LEFTCTRL: False,
    ecodes.KEY_LEFTALT: False,
    ecodes.KEY_LEFTMETA: False,
    ecodes.KEY_CAPSLOCK: False,
}

def check_special_keys(event):
    for special_key in special_keys_pressed:
        if event.code == special_key and (event.value == 1 or event.value == 2):
            special_keys_pressed[special_key] = True
        elif event.code == special_key and event.value == 0:
            special_keys_pressed[special_key] = False
    return False

toggle_keys_set = {
    ecodes.KEY_CAPSLOCK: False,
}

previous_key_down = 0

def check_toggle_keys(event):
    global previous_key_down
    if event.value == 0 and event.code == previous_key_down:
        for toggle_key in toggle_keys_set:
            if toggle_key == event.code:
                toggle_keys_set[toggle_key] = not toggle_keys_set[toggle_key]
                special_keys_pressed[toggle_key] = toggle_keys_set[toggle_key]
    else:
        previous_key_down = event.code
        
    return False

def kill_sequence(special_key, kill_key):
    def take_event(event):
        if special_keys_pressed[special_key] and event.code == kill_key:
            sys.exit()
        
        return False

    return take_event

def press(key):
    return lambda: virtual_keyboard.write(ecodes.EV_KEY, key, 1)

def unpress(key):
    return lambda: virtual_keyboard.write(ecodes.EV_KEY, key, 0)

def rule(event, source_key, output_events):
    if event.code == source_key:
        if event.value == 0 or event.value == 2:
            for output_event in output_events:
                output_event()

        return True
    else:
        return False

def special_key_rule(special_key, source_key, output_events, special_keys_not_pressed):
    def take_event(event):
        if special_keys_pressed[special_key] and not any([special_keys_pressed[x] for x in special_keys_not_pressed]):
            return rule(event, source_key, output_events)
        else:
            return False
    return take_event

def block_caps_lock(event):
    return event.code == ecodes.KEY_CAPSLOCK

def write_default_event(event):
    virtual_keyboard.write_event(event)
    return True

try:
    # when program is run from terminal, give terminal time to see the new line release
    time.sleep(1)

    keyboard = InputDevice('/dev/input/event0')
    keyboard.grab()

    virtual_keyboard = UInput()

    mappings = [
        check_special_keys,
        check_toggle_keys,
        # kill_sequence(ecodes.KEY_CAPSLOCK, ecodes.KEY_EQUAL),

        ## Inside text editor
        # text navigation
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_I, [press(ecodes.KEY_UP), unpress(ecodes.KEY_UP)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_J, [press(ecodes.KEY_LEFT), unpress(ecodes.KEY_LEFT)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_K, [press(ecodes.KEY_DOWN), unpress(ecodes.KEY_DOWN)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_L, [press(ecodes.KEY_RIGHT), unpress(ecodes.KEY_RIGHT)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),

        # pgUp/Down
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_U, [press(ecodes.KEY_PAGEUP), unpress(ecodes.KEY_PAGEUP)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_N, [press(ecodes.KEY_PAGEDOWN), unpress(ecodes.KEY_PAGEDOWN)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),

        # start/end line
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_A, [press(ecodes.KEY_HOME), unpress(ecodes.KEY_HOME)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_D, [press(ecodes.KEY_END), unpress(ecodes.KEY_END)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),

        # start/end file
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_W, [press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_HOME), unpress(ecodes.KEY_HOME), unpress(ecodes.KEY_LEFTCTRL)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_S, [press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_END), unpress(ecodes.KEY_END), unpress(ecodes.KEY_LEFTCTRL)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),

        # word navigation
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_H, [press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_LEFT), unpress(ecodes.KEY_LEFT), unpress(ecodes.KEY_LEFTCTRL)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_F, [press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_RIGHT), unpress(ecodes.KEY_RIGHT), unpress(ecodes.KEY_LEFTCTRL)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),

        ## Window navigation
        # tab navigation
        special_key_rule(ecodes.KEY_LEFTMETA, ecodes.KEY_J, [press(ecodes.KEY_LEFTCTRL), unpress(ecodes.KEY_LEFTMETA), press(ecodes.KEY_LEFTSHIFT), press(ecodes.KEY_TAB), unpress(ecodes.KEY_TAB), unpress(ecodes.KEY_LEFTSHIFT), unpress(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_LEFTMETA), press(ecodes.KEY_LEFTALT), unpress(ecodes.KEY_LEFTALT)], []),
        special_key_rule(ecodes.KEY_LEFTMETA, ecodes.KEY_L, [press(ecodes.KEY_LEFTCTRL), unpress(ecodes.KEY_LEFTMETA), press(ecodes.KEY_TAB), unpress(ecodes.KEY_TAB), unpress(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_LEFTMETA), press(ecodes.KEY_LEFTALT), unpress(ecodes.KEY_LEFTALT)], []),

        # workspace up/down
        special_key_rule(ecodes.KEY_LEFTMETA, ecodes.KEY_I, [press(ecodes.KEY_UP), unpress(ecodes.KEY_UP)], []),
        special_key_rule(ecodes.KEY_LEFTMETA, ecodes.KEY_K, [press(ecodes.KEY_DOWN), unpress(ecodes.KEY_DOWN)], []),

        # back/forward
        special_key_rule(ecodes.KEY_LEFTMETA, ecodes.KEY_A, [press(ecodes.KEY_LEFTALT), unpress(ecodes.KEY_LEFTMETA), press(ecodes.KEY_LEFT), unpress(ecodes.KEY_LEFT), unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTMETA), press(ecodes.KEY_LEFTALT), unpress(ecodes.KEY_LEFTALT)], []),
        special_key_rule(ecodes.KEY_LEFTMETA, ecodes.KEY_D, [press(ecodes.KEY_LEFTALT), unpress(ecodes.KEY_LEFTMETA), press(ecodes.KEY_RIGHT), unpress(ecodes.KEY_RIGHT), unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTMETA), press(ecodes.KEY_LEFTALT), unpress(ecodes.KEY_LEFTALT)], []),

        # map CapsLock-o to Alt-o(used as shortcut)
        special_key_rule(ecodes.KEY_CAPSLOCK, ecodes.KEY_O, [press(ecodes.KEY_LEFTALT), press(ecodes.KEY_O), unpress(ecodes.KEY_O), unpress(ecodes.KEY_LEFTALT)], [ecodes.KEY_LEFTCTRL, ecodes.KEY_LEFTALT, ecodes.KEY_LEFTMETA]),

        ## Terminal
        # keep common terminal ctrl-c, ctrl-l, ctrl-d, ctrl-x, ctrl-k, ctrl-u and ctrl-z
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_C, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_LEFTSHIFT), press(ecodes.KEY_C), unpress(ecodes.KEY_C), unpress(ecodes.KEY_LEFTSHIFT), unpress(ecodes.KEY_LEFTCTRL)], []),
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_L, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_L), unpress(ecodes.KEY_L), unpress(ecodes.KEY_LEFTCTRL)], []),
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_D, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_D), unpress(ecodes.KEY_D), unpress(ecodes.KEY_LEFTCTRL)], []),
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_X, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_X), unpress(ecodes.KEY_X), unpress(ecodes.KEY_LEFTCTRL)], []),
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_K, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_K), unpress(ecodes.KEY_K), unpress(ecodes.KEY_LEFTCTRL)], []),
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_U, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_U), unpress(ecodes.KEY_U), unpress(ecodes.KEY_LEFTCTRL)], []),
        special_key_rule(ecodes.KEY_LEFTALT, ecodes.KEY_Z, [unpress(ecodes.KEY_LEFTALT), press(ecodes.KEY_LEFTCTRL), press(ecodes.KEY_Z), unpress(ecodes.KEY_Z), unpress(ecodes.KEY_LEFTCTRL)], []),

        block_caps_lock,
        write_default_event
    ]

    for raw_event in keyboard.read_loop():
        if raw_event.type == ecodes.EV_KEY:
            event = special_keys_mapping(raw_event)
            for func in mappings:
                should_we_stop = func(event)
                if(should_we_stop):
                    break
            virtual_keyboard.syn()
finally:
    virtual_keyboard.close()
    keyboard.ungrab()

File: cua.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <dirent.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <libevdev/libevdev.h>
#include <libevdev/libevdev-uinput.h>

struct special_keys {
  int ctrl;
  int alt;
  int capslock;
  int meta;
};

struct input_device {
  int fd;
  struct libevdev *dev;
};

struct input_device get_input_device_by_name(char *input_device_name) {
  DIR *dir = opendir("/dev/input");
  struct dirent *de;
  
  while((de = readdir(dir)) != NULL) {
    if(de->d_type != DT_CHR) {
      continue;
    }
  	
    char dev_path[256];
    snprintf(dev_path, sizeof(dev_path), "/dev/input/%s", de->d_name);
    
    int fd = open(dev_path, O_RDONLY);
    if(fd < 0) {
      fprintf(stderr, "Couldn't open %s\n", dev_path);
      continue;
    }
    
    struct libevdev *dev;
    int err = libevdev_new_from_fd(fd, &dev);
    if(err) {
      close(fd);
      fprintf(stderr, "Libevdev input device error %s for %s\n", strerror(err), dev_path);
      continue;
    }
    
    const char *dev_name = libevdev_get_name(dev);
    const char *dev_phys = libevdev_get_phys(dev);
    
    if(!input_device_name) {
      printf("%s %s %s\n", dev_path, dev_name, dev_phys);
      
      // check phys ends with 0 as duplicate device names possible
    } else if(dev_name && dev_phys && strcmp(dev_name, input_device_name) == 0 && dev_phys[strlen(dev_phys) - 1] == '0') {
      closedir(dir);
      return (struct input_device){
        .fd = fd,
        .dev = dev
      };
    }
    
    libevdev_free(dev);
    close(fd);
  }
  
  closedir(dir);
  return (struct input_device){
    .fd = -1,
    .dev = NULL
  };
}

void press(struct libevdev_uinput *output_dev, int code, int value) {
  libevdev_uinput_write_event(output_dev, EV_KEY, code, value);
}

void press_mod1(struct libevdev_uinput *output_dev, int mod, int code, int value) {
  if(value == 1) {
    press(output_dev, mod, 1);
    press(output_dev, code, 1);
  } else if(value == 2) {
    // don't repeat mod key as may get stuck
    press(output_dev, code, 2);
  } else if(value == 0) {
    press(output_dev, code, 0);
    press(output_dev, mod, 0);
  }
}

void press_mod2(struct libevdev_uinput *output_dev, int mod1, int mod2, int code, int value) {
  if(value == 1) {
    press(output_dev, mod1, 1);
    press(output_dev, mod2, 1);
    press(output_dev, code, 1);
  } else if(value == 2) {
    // don't repeat mod key as may get stuck
    press(output_dev, code, 2);
  } else if(value == 0) {
    press(output_dev, code, 0);
    press(output_dev, mod1, 0);
    press(output_dev, mod2, 0);
  }
}

int main(int argc, char **argv) {
  if(argc < 2) {
    get_input_device_by_name(NULL);
    fprintf(stderr, "Missing device name\n");
    return 1;
  }

  char *input_device_name = argv[1];
  
  struct input_device input = { .fd = -1, .dev = NULL };
  struct libevdev_uinput *output_dev = NULL;
  
  for(;;) {
    if(!input.dev) {
      input = get_input_device_by_name(input_device_name);
      if(!input.dev) {
        sleep(3);
        continue;
      }
    }

    int uifd = open("/dev/uinput", O_RDWR);
    if(uifd < 0) {
      fprintf(stderr, "Couldn't open /dev/uinput\n");
      return 1;
    }

    int err = libevdev_uinput_create_from_device(input.dev, uifd, &output_dev);
    if(err) {
      fprintf(stderr, "Libevdev output device error %s for %s\n", strerror(err), input_device_name);
      return 1;
    }

    err = libevdev_grab(input.dev, LIBEVDEV_GRAB);
    if(err) {
      fprintf(stderr, "Libevdev grab error %s for %s\n", strerror(err), input_device_name);
      return 1;
    }

    struct input_event event;
    struct special_keys spk = {
      .ctrl = 0,
      .alt = 0,
      .capslock = 0,
      .meta = 0,
    };

    int previous_key = 0;
    int capslock_toggle = 0;

    for(;;) {

    int rc = libevdev_next_event(input.dev, LIBEVDEV_READ_FLAG_NORMAL | LIBEVDEV_READ_FLAG_BLOCKING, &event);
    if(rc == -ENODEV) {
      libevdev_uinput_destroy(output_dev);
      output_dev = NULL;
      libevdev_free(input.dev);
      close(input.fd);
      close(uifd);
      input = (struct input_device){ .fd = -1, .dev = NULL };
      sleep(1);
      break;
    }

    if(event.type != EV_KEY) {
      continue;
    }    

    const char *event_name = libevdev_event_code_get_name(event.type, event.code);
    // printf("Event: %s %d\n", event_name, event.value);

    // switch ctrl and alt
    if(event.code == KEY_LEFTALT) {
      spk.alt = event.value;
      press(output_dev, KEY_LEFTCTRL, event.value);
    } else if(event.code == KEY_LEFTCTRL) {
      spk.ctrl = event.value;

    } else if(event.code == KEY_CAPSLOCK) {
      // use capslock and meta as modifier keys
      if(event.value == 1) {
        spk.capslock = 1;
      } else if(event.value == 0) {
        if(previous_key == KEY_CAPSLOCK) {
          // capslock is sticky
          if(capslock_toggle) {
            capslock_toggle = 0;
            spk.capslock = 0;
          } else {
            capslock_toggle = 1;
            spk.capslock = 1;
          }
        } else {
          spk.capslock = 0;
          capslock_toggle = 0;
        }
      } 
    } else if(event.code == KEY_LEFTMETA) {
      spk.meta = event.value;
      if(previous_key == KEY_LEFTMETA && event.value == 0) {
        press(output_dev, KEY_LEFTMETA, 1);
        press(output_dev, KEY_LEFTMETA, 0);
      }

    } else if(spk.capslock && !spk.ctrl && !spk.alt && !spk.meta) {
      switch(event.code) {
        // basic navigation
        case KEY_I: press(output_dev, KEY_UP, event.value); break;
        case KEY_J: press(output_dev, KEY_LEFT, event.value); break;
        case KEY_K: press(output_dev, KEY_DOWN, event.value); break;
        case KEY_L: press(output_dev, KEY_RIGHT, event.value); break;
        // pgUp/pgDown
        case KEY_U: press(output_dev, KEY_PAGEUP, event.value); break;
        case KEY_N: press(output_dev, KEY_PAGEDOWN, event.value); break;
        // start and end of line
        case KEY_A: press(output_dev, KEY_HOME, event.value); break;
        case KEY_D: press(output_dev, KEY_END, event.value); break;
        // start and end of file
        case KEY_W: press_mod1(output_dev, KEY_LEFTCTRL, KEY_HOME, event.value); break;
        case KEY_S: press_mod1(output_dev, KEY_LEFTCTRL, KEY_END, event.value); break;
        // start and end of word
        case KEY_H: press_mod1(output_dev, KEY_LEFTCTRL, KEY_LEFT, event.value); break;
        case KEY_F: press_mod1(output_dev, KEY_LEFTCTRL, KEY_RIGHT, event.value); break;
        // autophrase
        case KEY_3: press(output_dev, KEY_A, event.value); break;

        default: press(output_dev, event.code, event.value);
      }
    } else if(spk.ctrl && spk.meta && event.code == KEY_L) {
      // lock screen
      press_mod1(output_dev, KEY_LEFTMETA, KEY_L, event.value);
    } else if(spk.meta) {
      switch(event.code) {
        // tab left and right
        case KEY_J: press_mod2(output_dev, KEY_LEFTCTRL, KEY_LEFTSHIFT, KEY_TAB, event.value); break;
        case KEY_LEFT: press_mod2(output_dev, KEY_LEFTCTRL, KEY_LEFTSHIFT, KEY_TAB, event.value); break;
        case KEY_L: press_mod1(output_dev, KEY_LEFTCTRL, KEY_TAB, event.value); break;
        case KEY_RIGHT: press_mod1(output_dev, KEY_LEFTCTRL, KEY_TAB, event.value); break;
        // workspace up and down
        case KEY_I: press_mod1(output_dev, KEY_LEFTMETA, KEY_UP, event.value); break;
        case KEY_UP: press_mod1(output_dev, KEY_LEFTMETA, KEY_UP, event.value); break;
        case KEY_K: press_mod1(output_dev, KEY_LEFTMETA, KEY_DOWN, event.value); break;
        case KEY_DOWN: press_mod1(output_dev, KEY_LEFTMETA, KEY_DOWN, event.value); break;
        // fullscreen
        case KEY_F: press_mod1(output_dev, KEY_LEFTMETA, KEY_F, event.value); break;

        default: press(output_dev, event.code, event.value);
      }
    } else if(spk.ctrl) {
      switch(event.code) {
        // ctrl-c, ctrl-l, ctrl-d, ctrl-x, ctrl-k, ctrl-u, ctrl-z
        case KEY_C: press_mod2(output_dev, KEY_LEFTCTRL, KEY_LEFTSHIFT, KEY_C, event.value); break;
        case KEY_L: press_mod1(output_dev, KEY_LEFTCTRL, KEY_L, event.value); break;
        case KEY_D: press_mod1(output_dev, KEY_LEFTCTRL, KEY_D, event.value); break;
        case KEY_X: press_mod1(output_dev, KEY_LEFTCTRL, KEY_X, event.value); break;
        case KEY_K: press_mod1(output_dev, KEY_LEFTCTRL, KEY_K, event.value); break;
        case KEY_U: press_mod1(output_dev, KEY_LEFTCTRL, KEY_U, event.value); break;
        case KEY_Z: press_mod1(output_dev, KEY_LEFTCTRL, KEY_Z, event.value); break;

        default: press(output_dev, event.code, event.value);
      }
    } else {
      // we didn't match any rules, pass along event
      press(output_dev, event.code, event.value);
    }

    previous_key = event.code;

    libevdev_uinput_write_event(output_dev, EV_SYN, SYN_REPORT, 0);
    }
  }
}
